x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/repositories/DeleteCcrRestoreSessionAction.java

```diff
@@ -0,0 +1,88 @@
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.shard.ShardNotFoundException;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.ccr.repository.CcrRestoreSourceService;
+
+public class DeleteCcrRestoreSessionAction extends Action<DeleteCcrRestoreSessionAction.DeleteCcrRestoreSessionResponse> {
+
+    public static final DeleteCcrRestoreSessionAction INSTANCE = new DeleteCcrRestoreSessionAction();
+    private static final String NAME = "internal:admin/ccr/restore/session/delete";
+
+    private DeleteCcrRestoreSessionAction() {
+        super(NAME);
+    }
+
+    @Override
+    public DeleteCcrRestoreSessionResponse newResponse() {
+        return new DeleteCcrRestoreSessionResponse();
+    }
+
+    public static class TransportDeleteCcrRestoreSessionAction
+        extends TransportSingleShardAction<DeleteCcrRestoreSessionRequest, DeleteCcrRestoreSessionResponse> {
+
+        private final IndicesService indicesService;
+        private final CcrRestoreSourceService ccrRestoreService;
+
+        @Inject
+        public TransportDeleteCcrRestoreSessionAction(ThreadPool threadPool, ClusterService clusterService, ActionFilters actionFilters,
+                                                      IndexNameExpressionResolver resolver, TransportService transportService,
+                                                      IndicesService indicesService, CcrRestoreSourceService ccrRestoreService) {
+            super(NAME, threadPool, clusterService, transportService, actionFilters, resolver, DeleteCcrRestoreSessionRequest::new,
+                ThreadPool.Names.GENERIC);
+            this.indicesService = indicesService;
+            this.ccrRestoreService = ccrRestoreService;
+        }
+
+        @Override
+        protected DeleteCcrRestoreSessionResponse shardOperation(DeleteCcrRestoreSessionRequest request, ShardId shardId) {
+            IndexShard indexShard = indicesService.getShardOrNull(shardId);
+            if (indexShard == null) {
+                throw new ShardNotFoundException(shardId);
+            }
+            ccrRestoreService.closeSession(request.getSessionUUID(), indexShard);
+            return new DeleteCcrRestoreSessionResponse();
+        }
+
+        @Override
+        protected DeleteCcrRestoreSessionResponse newResponse() {
+            return new DeleteCcrRestoreSessionResponse();
+        }
+
+        @Override
+        protected boolean resolveIndex(DeleteCcrRestoreSessionRequest request) {
+            return false;
+        }
+
+        @Override
+        protected ShardsIterator shards(ClusterState state, InternalRequest request) {
+            final ShardId shardId = request.request().getShardId();
+            // The index uuid is not correct if we restore with a rename
```

Comment:
I've added a comment to CcrRepository how we can ensure we have the correct shard id.
we should have no leniency here anymore then, but do an actual lookup by shard id.

Reply:
ClearCcrRestoreSessionAction is no longer a shard action. It is now `TransportNodesAction`. I made this change because the session will be on the node that started the session.